### PR TITLE
player rb [wip]

### PR DESCRIPTION
**Pull request by CMorooney**
I have to run to an event, hoping to look at this some more tomorrow, but jic you notice something I don't I'm tossing up this branch

right now there's weird problem where `moveWithCollisions` in `RigidBody.update` will return 0 collision every other call while the player is standing still on the ground, which flips `onGround` for the player RB rapidly which makes everything very buggy and difficult to work on -- notice anything I am missing?

### CODE REVIEW

**Comment by CMorooney** [comment]
is `main` the wrong destination branch? feel free to change

**Comment by Franchovy** [comment]
> I have to run to an event, hoping to look at this some more tomorrow, but jic you notice something I don't I'm tossing up this branch
> 
> right now there's weird problem where `moveWithCollisions` in `RigidBody.update` will return 0 collision every other call while the player is standing still on the ground, which flips `onGround` for the player RB rapidly which makes everything very buggy and difficult to work on -- notice anything I am missing?

I'll take a look. Typically it's the y-velocity being 0 that triggers "no collisions" so the onGround turns off, then it switches to falling which collides with the ground. Will take a look now

**Comment by Franchovy** [comment]
Issue's fixed. I left a comment to explain the new logic

**Comment by Franchovy** [comment]
@CMorooney You'll have to merge main into this so you have all the latest updates!

File: src/sprites/RigidBody.lua [hunk 1]
```diff
@@ -88,7 +88,16 @@ function RigidBody:update()
 
   -- incorporate gravity
 
-  self.velocity = self.velocity + (gmt.vector2D.new(0, 1) * _G.delta_time) * self.g_mult
+  if not groundFound then
+    -- Adds gravity vector to current velocity
+
+    self.velocity = self.velocity + (gmt.vector2D.new(0, 1) * _G.delta_time) * self.g_mult
```
Comment (Franchovy) [inline src/sprites/RigidBody.lua:113]: @CMorooney if we want to cap max fall speed let's add the following lines in here. Something like this:

if self.velocity.y > [maxFallSpeed] then 
    self.velocity.y = maxFallSpeed
end

**Comment by CMorooney** [comment]
@Franchovy still having some issues jumping off of a vertically moving platform. Planning on looking at it this weekend but somehow ended up with _another_ goat emergency. Idk why everyone is trying (and succeeding) to die so hard on the farm this summer

will have time tomorrow as well but if you want to take a look feel free

**Comment by Franchovy** [comment]
Modified the elevator behavior:

- Now works through *continuous* activation. You have to maintain the button press to move in the given direction.
- Also handles moving in the opposite direction, up to the starting point.
- You can now set the starting point to be different from the initial starting point through LDtk (didn't include an example though)

We still have things to fix:
- If touching the elevator on the side, the activation call still happens. The only position that should activate the elevator is standing on top of it (a simple normal check should handle this), but for horizontal movement, should only activate if the player position is >= the center X of the elevator (or <= if moving in the -x direction).
- If standing on top of the elevator, the player movement should be nullified if able to move in the same direction as the elevator. 
To support this, we return `true` / `false` from the call to elevator:activate() to indicate whether the elevator captured the player movement or not. We then need this boolean to cancel horizontal movement from the player if returned as true.

**Comment by Franchovy** [comment]
Hope the animals are doing ok <3

**Comment by Franchovy** [comment]
Pushed an update:
- Player class `shouldExitParent` is a variable that can be set to true from anywhere and will reset on next call from RigidBody `exitParent()`.
- This fixed most of the behaviors.

We have some awkwardness in the behavior of the elevators but they are now fully useable in my opinion. Please review and let me know if you agree.

The small points of improvement are: 
- Automatically teleporting to the center of the elevator (should only happen on successful activation)
- Bumping into the left or right of the elevator creates very small -y displacement.

**Comment by Franchovy** [comment]
Fixed the first improvement point.

**Review by Franchovy — APPROVED** [review]

**Comment by CMorooney** [comment]
lost a goat :( had to go find more so the other didn't lose her mind, not sure how to save my own 🙃 

I am having trouble with vertical elevator in the physics test level so I might try and take a look at that soon, seems like I can't make an 'up' elevator -- using a positive `distance` allows me to move down, but a negative distance jumps the elevator over the player when trying to go up

**Comment by Franchovy** [comment]
Realized I made a mistake updating the `self.displacement` inside the `activate()` function of the Elevator. It needs to take into account the `_G.delta_time` in order to properly calculate how much displacement is left.

**Pull request merged**